Unit Testing Overhaul

mmorrise opened the pull request:
See #1337.
This pull request includes the following new/updated files:
- [x] Testing LaTeX file
- [x] Figures folder with pictures for the LaTeX file
- [x] solutions.py and test_solutions.py - Lab specs for the students
- [x] solutions_ANSWER.py and test_solutions_ANSWER.py - Answer keys for the spec files
- [x] hands folder containing test cases for the test_solutions_ANSWER.py file

This pull request does not contain a test driver. How this lab should be graded has not yet been decided. Here are my suggestions on this:
- Run `py.test --cov` to on the student's code to ensure that all their test cases pass, and they have full coverage of the functions they are testing
- Review sections of the students code to manually see the thoroughness of test functions (for the problem on the Set game).
- Write some unit tests which should pass of the code that the student has written (for the last problem) and verify that these tests pass. (To do this, the function names, input and output would need to be specified for the students in advance, which is currently not the case).

At the moment, Problems 4 and 5, which have to do with Test Driven Design, are not well polished. I would most appreciate feedback in this area.

shanemcq18 commented:
@mariafabiano and @cassief2 could you proofread this for us and give feedback? Here's what you need to do to get these changes locally:
```bash
(start on develop branch)
git remote add martha https://github.com/mmorrise/Lab-Development
git checkout -b martha_testing
git pull martha develop
```
If you have changes to make to the pull request, commit your changes while on your `martha_testing` branch and do `git push martha develop`. I hope this works, but it might problematic since the PR branch is `develop` (@mmorrise, next time please work on a branch other than `develop`).

Whether or not you have changes, would you comment here with your feedback? We need to get this finished ASAP.

@mmorrise I will also have feedback for you soon.

cassief2 commented:
I know Shane showed how to turn off the blue coloring of some words in the yellow text boxes...but I don't remember how to do it.  Fix the yellow boxes where you run py.test so no words turn blue.

mariafabiano commented:
Hey @shanemcq18, I made some minor edits to the TeX file (grammar, rewording/clarifying, etc.). While on `martha_testing` I tried pushing the changes, but I got this error:
`remote: Permission to mmorrise/Lab-Development.git denied to mariafabiano.`
`fatal: unable to access 'https://github.com/mmorrise/Lab-Development/': The requested URL returned error: 403`

Overall, I think it looks great. I enjoyed doing the lab. A couple of thoughts:
- To use the `tree` command, I had to install HomeBrew and then install the `tree` package before I could get it to work
- Problem 1: some of my addition tests failed because I tried floats; it might be worthwhile to clarify we're doing integer operations (similar issues with Problems 2 and 3)

shanemcq18 commented:
To turn off coloring, surround the text in the lstlisting with `<<` and `>>`.

@mariafabiano and @cassief2, I guess you can't push to this branch (which is okay, I just wanted to see if it would work). Please expand your comments listed above if you have more feedback. @mmorrise, go ahead and make changes to reflect the feedback listed above.

shanemcq18 commented on Vol1B/Testing/Testing.tex:
```diff
@@ -1,41 +1,39 @@
 \lab{Unit Testing In Python}{Unit Testing In Python}
 
-\objective{Convey the importance of unit testing in code, especially in the corporate world. Unit testing in python is easy, valuable, and powerful. Any project manager will appreciate a developer who is able to accurately test their code.}
+\objective{Convey the importance of unit testing in code, especially in the corporate world. Unit testing in python is easy, valuable, and powerful. Help students understand how they can use unit testing to become more effective and productive programmers.}
```
The objective should address the student, not the teacher or the lab author (so saying "help students understand" makes no sense). Start right off the bat by talking about what testing is and why it's important (for everyone, not just software developers). Much of the first paragraph or two could be cut up and turned into a compelling objective. Eliminate the wordiness in the first few paragraphs though: make it succinct and professional, but not robotic.

shanemcq18 commented on Vol1B/Testing/Testing.tex:
```diff
@@ -1,41 +1,39 @@
 \section*{PyTest}
 
-Okay, so we've disscussed enough of the why. Now we will explore the "how" of testing code in python.
+Now that we have discussed the ``why" of unit testing, we will discuss the ``how" of testing code in Python.
```
Sentences like this are pointless in the lab. Don't talk about what you're going to talk about, just start talking about it.

shanemcq18 commented on Vol1B/Testing/Testing.tex:
```diff
@@ -200,10 +190,10 @@ \section*{pytest as a module}
 \end{problem}
 
-Pytest also offers an option to make "pytest fixtures". These are wildly useful for mocking data or preventing constant code duplication.
+Suppose you want to use the same pieces of code in different test functions. Pytest has a feature called "fixtures" for mocking data that reduces code duplication. 
```
Instead of putting words in quotes ("fixtures"), italicize them with `\emph{}` (`\emph{fixtures}` spits out _fixtures_).

mmorrise commented:
I have made changes based on the feedback I received. Let me know if there is anymore changes to make.

shanemcq18 commented:
@mmorrise I made a few tweaks to a few spots of the lab (nothing major, just a few changes to the wording). Use `git pull origin testing` to get my changes. I want you to change one thing before we merge this pull request: in the first problem on Set, we say
> Each card in Set can be uniquely represented by a 4 bit integer in base 3, where each digit represents a different property and each property has 3 possible values. Represent a group of Set cards as a text file where each line contains a 4 bit number representing one card.

Can we give them an example of what this means? That is, that the numbers `0000` and `1210` represent cards, but `000` and `0123` do not (and say why). Be VERY EXPLICIT in as you explain this, as if the reader has no idea what a 4-bit integer in base 3 is (because some readers won't!).

mmorrise commented:
I have added the clarification on the "4 bit number in base 3" as a foot note. I also changed the wording of part of the problem slightly. I think it would be good if the TAs also went over this in lab.

I did not add a note about how "test" must be in the file name and function name, since this is already in the text, and I think it is sufficient as is.
